File: dcp-solutions/2025-01-20/solution.js
### JavaScript Solution
```markdown
class HitCounter {
    constructor() {
        this.cache = {};
    }

    record(timestamp) {
        if (!(timestamp in this.cache)) {
            this.cache[timestamp] = 1;
        } else {
            this.cache[timestamp]++;
        }
    }

    total() {
        return Object.values(this.cache).reduce((a, b) => a + b, 0);
    }

    range(lower, upper) {
        let count = 0;

        for (let ts in this.cache) {
            if (ts >= lower && ts <= upper) {
                count += this.cache[ts];
            } else if (ts > upper) break;
        }

        return count;
    }
}
```

File: dcp-solutions/2025-01-20/solution.rb
### Ruby Solution

```markdown
class HitCounter
  def initialize()
    @map = {}
  end

  def record(timestamp)
    @map[timestamp] ||= [0]
    @map[timestamp][0] += 1
  end

  def total
    @map.values.sum
  end

  def range(lower, upper)
    count = 0
    i = lower
    while i <= upper do
      count += @map[i] || 0
      i += 1
    end
    count
  end
end
```